### PR TITLE
Implemented contains for BoundingBox containing other BoundingBox

### LINKED ISSUE #2896: Implement contains for BoundingBox
## Description
Implement `__contains__` for `BoundingBox` containing either a point, or another `BoundingBox`. This means that users could then write:

`if point in box:` or `if little_box in big_box`.

## Alternatives
It is possible for users to accomplish this currently but requires some clever coding to avoid becoming difficult to read:

``` python
def in_box(point, box):
   for min_p, p, max_p in zip(box.lower_left, point, box.upper_right):
     if p < min_p or > max_p:
        return False 
   return True
```

## Compatibility
This would be an enhancement, and would not alter the behavior of the existing API. 

There is a risk though that users will misinterpret the results. A point in the bounding box of a volume *may* be in the volume, but not necessarily. A user could misuse this information and create problems for themselves. Also a small volume's bounding box can be completely contained in another volume's bounding box, and be completely outside that other volume.

### PR DESCRIPTION
**Pull request by MicahGale**
# Description

This is a quality of life feature request that tests if things are inside of a `BoundingBox`. Originally the goal was to have `if point in box` and `if other_box in box`. Turns out since #2759, `if point in box` has been implemented.

This PR does the following:

* Implement checking if a bounding box is inside of another: `if other_box in box`
* Implements length enforcement for points so: `[1]*100 in box` gives a helpful error.
* Implements type checking so `if "foo" in box` gives a helpful error.
* Adds two unit tests to test the `__contains__` function for `BoundingBox` to test it's accuracy, and the other tests it's error handling. 

It is unclear if this should be explicitly documented for users or not. 

Fixes #2896 

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

**Comment by MicahGale** [comment]
Another PR (#2912) the CI is magically working again? Could we try restarting these runs to see if they work @pshriwise?

File: tests/unit_tests/test_bounding_box.py [hunk 1]
```diff
@@ -156,3 +156,27 @@ def test_bounding_box_methods():
...
+        (test_bb_1, openmc.BoundingBox((-9, -19, -29), (0, 0, 0)), True)
+    ],
+)
+def test_bounding_box_contains(bb, other, expected):
+    assert (other in bb) == expected
+
+@pytest.mark.parametrize(
+    "invalid, ex",
+    [
+        ((1,0), ValueError),
```
Comment (shimwell) [inline tests/unit_tests/test_bounding_box.py:175]:
```suggestion
        ((1, 0), ValueError),
```

File: tests/unit_tests/test_bounding_box.py [hunk 2]
```diff
@@ -156,3 +156,27 @@ def test_bounding_box_methods():
...
+        ((1,2,3,4), ValueError),
```
Comment (shimwell) [inline tests/unit_tests/test_bounding_box.py:176]:
```suggestion
        ((1, 2, 3, 4), ValueError),
```

File: tests/unit_tests/test_bounding_box.py [hunk 3]
```diff
@@ -45,7 +45,7 @@ def test_bounding_upper_right(bb, expected):
 
 
 @pytest.mark.parametrize(
-    "bb, expected",
+    "bb, expected",
```
Comment (shimwell) [inline tests/unit_tests/test_bounding_box.py:48]:
```suggestion
    "bb, expected",
```

File: openmc/bounding_box.py [hunk 4]
```diff
@@ -95,9 +95,25 @@ def __or__(self, other: BoundingBox) -> BoundingBox:
...
         return new
 
-    def __contains__(self, point):
-        """Check whether or not a point is in the bounding box"""
-        return all(point > self.lower_left) and all(point < self.upper_right)
+    def __contains__(self, other):
+        """Check whether or not a point or another bounding box is in the bounding box.
+
+        For another bounding box to be in the parent it must lie fully inside of it,
+        with no overlap.
```
Comment (shimwell) [inline openmc/bounding_box.py:102]: I'm not sure if overlap is the right word here.

**Comment by shimwell** [comment]
This looks like a useful feature to me. I've made a few comments for consideration but nothing major

I would recommend one more test that checks a bounding box is in another bounding box when the first bounding box is only partly within. This would return false.

File: openmc/bounding_box.py (hunk 4 above)
Comment (MicahGale) [reply openmc/bounding_box.py:102]: Intersect?

File: openmc/bounding_box.py [hunk 5]
```diff
@@ -95,9 +95,25 @@ def __or__(self, other: BoundingBox) -> BoundingBox:
...
+        # test for a single point
+        if isinstance(other, (tuple, list, np.ndarray)):
+            point = other
+            check_length("Point", point, 3, 3)
+            return all(point > self.lower_left) and all(point < self.upper_right)
+        elif isinstance(other, BoundingBox):
+            return all([p in self for p in [other.lower_left, other.upper_right]])
+        else:
+            raise TypeError(
+                f"Unable to test if {other} is in the bounding box."
```
[the remark this answers is not in the document]
Comment (MicahGale) [reply openmc/bounding_box.py:114]: ascertain seems almost too formal. I'm thinking "determine"?

File: openmc/bounding_box.py (hunk 4 above)
Comment (pshriwise) [reply openmc/bounding_box.py:102]: Nit-pick alert: I think the final phrase can simply be removed to make the statement less ambiguous.

Comment (pshriwise) [reply openmc/bounding_box.py:102]: Yes, exactly

**Comment by MicahGale** [comment]
Ok that test was added. I also added one for one of the vertices being coincident because the objects were right there.

**Review by shimwell — APPROVED** [review]
A nice PR that improves on the existing __contains__ method to supports bounding boxes, includes tests, doc strings updates, pep8 formatting. All looks good to me and I'm happy to merge this one in. Many thanks for this improved method

Will merge Tuesday afternoon if there are no objections.

**Comment by MicahGale** [comment]
Thanks @shimwell.

**Review by paulromano — CHANGES_REQUESTED** [review]
I'd like to take a look at this before we merge. Sorry to delay! Will get a review in asap

**Review by paulromano — APPROVED** [review]

**Pull request merged**